[LV] Optimize VPWidenIntOrFpInductionRecipe for known TC

hazzlim opened the pull request:
Optimize the IR generated for a VPWidenIntOrFpInductionRecipe to use the
narrowest type necessary, when the trip-count of a loop is known to be
constant and the only use of the recipe is the condition used by the
vector loop's backedge branch.

hazzlim committed:
[LV] Pre-commit tests for optimizing induction variable width

hazzlim committed:
[LV] Optimize VPWidenIntOrFpInductionRecipe for known TC

Optimize the IR generated for a VPWidenIntOrFpInductionRecipe to use the
narrowest type necessary, when the trip-count of a loop is known to be
constant and the only use of the recipe is the condition used by the
vector loop's backedge branch.

hazzlim commented:
I wasn't sure if this was the best/right way to approach this, so please let me know if there's a better approach to take!

hazzlim commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2150,6 +2150,9 @@ class VPWidenIntOrFpInductionRecipe : public VPHeaderPHIRecipe {
   VPValue *getStepValue() { return getOperand(1); }
   const VPValue *getStepValue() const { return getOperand(1); }
 
+  /// Update the start value of the recipe.
```
Ah yep good point - fixed.

hazzlim commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -661,6 +661,60 @@ static void recursivelyDeleteDeadRecipes(VPValue *V) {
+    return;
+
+  // Calculate the widest type required for known TC and VF.
+  uint64_t Width = BestVF.getKnownMinValue();
```
Ah yes - I have updated to multiply by BestUF here as well.

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.h:
```diff
@@ -59,8 +59,9 @@ struct VPlanTransforms {
-  /// Optimize \p Plan based on \p TC and \p BestVF.
-  static void optimizeForTCAndVF(VPlan &Plan, unsigned TC, ElementCount BestVF);
+  /// Optimize \p Plan based on \p TC, \p BestVF and \p BestUF.
+  static void optimizeForTCAndVFAndUF(VPlan &Plan, unsigned TC,
```
nit: Perhaps just call it `optimizeForTCAndVFUF`

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -661,14 +661,16 @@ static void recursivelyDeleteDeadRecipes(VPValue *V) {
-  // Calculate the widest type required for known TC and VF.
-  uint64_t Width = BestVF.getKnownMinValue();
+  // Calculate the widest type required for known TC, VF and UF.
+  uint64_t Width = BestVF.getKnownMinValue() * BestUF;
```
After making this change I didn't see any corresponding change in the tests, which looks we have a missing test case. Is it possible to create a very simple test that is affected by this change? For example, if you have a known trip count of 50 and use `-force-vector-width=64 -force-vector-interleave=8` then the max value will be 511 and will require an i16 type. That's still better than an i64 type though!

hazzlim replied:
Ah good point. I've added a test - but this PR will still not trigger on the case when we are interleaving. This is because when unrolling the induction by UF, we create new recipes for the induction steps for part 1 .. UF, which are users of the original IndVar. As this PR limits to only a single use (Cmp) of the IndVar, this case will not be optimized.
https://github.com/llvm/llvm-project/blob/fbc18b85d6ce5ab6489a2b08f9b38d446fe9d6f6/llvm/lib/Transforms/Vectorize/VPlanUnroll.cpp#L186-L218

I am thinking we could leave extending this to a follow-up PR, if you agree?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7657,6 +7657,9 @@ DenseMap<const SCEV *, Value *> LoopVectorizationPlanner::executePlan(
   VPlanTransforms::optimizeForVFAndUF(BestVPlan, BestVF, BestUF, PSE);
 
+  auto TC = PSE.getSE()->getSmallConstantTripCount(OrigLoop);
+  VPlanTransforms::optimizeForTCAndVF(BestVPlan, TC, BestVF);
```
Can we instead use BestVPlan.getTripCount() and move this in optimizeForVFAndUF, as this also optimizes the plan for a given UF/VF?

david-arm replied:
Hi @fhahn, I thought that you were planning to move optimizeForVFAndUF at some point so that it's called for every VF and UF during the planning stage, rather than after we've chosen the best VF and UF? If @hazzlim moves the changes into `optimizeForVFAndUF` would that interfere with that?

fhahn replied:
I don't think it would interfere with that, in general I assume it might make sense to apply this one for all potential VFs as narrower IVs might trip the balance eventually (at the moment, the cost of inductions isn't estimate correctly yet though).

hazzlim replied:
I've moved into optimizeForVFAndUF, splitting out the existing code and this new code into static functions as it seemed sensible to make the logic simpler/things easier to read, if you agree?

fhahn commented on llvm/test/Transforms/LoopVectorize/tail-folding-optimize-vector-induction-width.ll:
```diff
@@ -0,0 +1,898 @@
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32"
+target triple = "aarch64-unknown-linux-gnu"
+
+define dso_local void @canonical_small_tc_i8(ptr nocapture noundef writeonly %p) {
```
```suggestion
define void @canonical_small_tc_i8(ptr nocapture noundef writeonly %p) {
```

hazzlim replied:
Thanks - done.

hazzlim commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -661,6 +661,60 @@ static void recursivelyDeleteDeadRecipes(VPValue *V) {
+        Bound != Plan.getOrCreateBackedgeTakenCount())
+      continue;
+
+    if (NewIVTy == WideIV->getScalarType())
```
Thanks - done.

hazzlim commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -661,6 +661,64 @@ static void recursivelyDeleteDeadRecipes(VPValue *V) {
+    // comparison with the backedge-taken-count.
+    VPValue *Bound;
+    using namespace VPlanPatternMatch;
+    auto *Cmp = dyn_cast<VPInstruction>(*WideIV->user_begin());
```
Thanks - done.

hazzlim commented on llvm/test/Transforms/LoopVectorize/tail-folding-optimize-vector-induction-width.ll:
```diff
@@ -0,0 +1,898 @@
+; RUN: opt -p loop-vectorize -prefer-predicate-over-epilogue=predicate-dont-vectorize -S %s | FileCheck %s
+
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32"
+target triple = "aarch64-unknown-linux-gnu"
```
Good point - done and removed the target datalayout as well, as this seems unnecessary also.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -661,11 +663,63 @@ static void recursivelyDeleteDeadRecipes(VPValue *V) {
+    return false;
+
+  // Calculate the widest type required for known TC, VF and UF.
+  uint64_t TCVal = TC->getZExtValue();
```
This assumes the trip count first into 64 bits, which may not be the case I think?

hazzlim replied:
Ah yep good point - fixed.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -661,11 +663,63 @@ static void recursivelyDeleteDeadRecipes(VPValue *V) {
+    if (!match(*WideIV->user_begin(),
+               m_Binary<Instruction::ICmp>(m_Specific(WideIV),
+                                           m_VPValue(Bound))) ||
+        Bound != Plan.getOrCreateBackedgeTakenCount())
```
Can we also use `m_SPecific` for the bound?

hazzlim replied:
Thanks, that's much neater  - Done.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -661,11 +663,65 @@ static void recursivelyDeleteDeadRecipes(VPValue *V) {
+  if (!TC || !BestVF.isFixed())
+    return false;
+
+  auto TCVal = TC->getValue().tryZExtValue();
+  if (!TCVal)
+    return false;
```
Can we instead do the math on APInt to not require this restriction?

Could you also add a test with TC > 64 bits?

hazzlim replied:
Sure - I've updated to do the majority of the calculations on APInt to avoid issues with bitwidth, and I've added a test with TC > 64 bits :)

hazzlim commented:
@fhahn polite ping on this - the PR had gone a bit out of sync with recent updates to `optimizeForTCAndVFAndUF`, so I've rebased & fixed things.

david-arm commented on llvm/test/Transforms/LoopVectorize/tail-folding-optimize-vector-induction-width-unrolling.ll:
Given how large the file is, and knowing the optimisation doesn't even trigger perhaps it's worth removing the test for now?

david-arm reviewed (commented):
I think this looks almost ready to go! I just had two more minor comments ...

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -975,11 +978,74 @@ void VPlanTransforms::simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy) {
+    auto AlignedTC =
+        Align * APIntOps::RoundingUDiv(TC, APInt(TC.getBitWidth(), Align),
+                                       APInt::Rounding::UP);
+    auto MaxVal = AlignedTC - 1;
```
This is an APInt? Would probably be clearer to spell it out.
```suggestion
    APInt MaxVal = AlignedTC - 1;
```

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -975,11 +978,74 @@ void VPlanTransforms::simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy) {
+
+  // Calculate the widest type required for known TC, VF and UF.
+  auto ComputeBitWidth = [](APInt TC, uint64_t Align) {
+    auto AlignedTC =
```
APInt?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -975,11 +978,74 @@ void VPlanTransforms::simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy) {
+
+  VPBasicBlock *HeaderVPBB = Plan.getVectorLoopRegion()->getEntryBasicBlock();
+  for (VPRecipeBase &Phi : HeaderVPBB->phis()) {
+    auto *WideIV = dyn_cast<VPWidenIntOrFpInductionRecipe>(&Phi);
```
Can you add a comment saying why this is restricted to canonical IVs with a single user?

hazzlim replied:
Done.

fhahn reviewed (commented):
One more thing I am wondering if we could end up introducing illegal vector types for a target and if we have to check the new type is legal?

hazzlim committed:
Address Review Comments

- Use cast rather than dyn_cast
- Remove unrolling test
- Be explicit about APInt type.
- Add comment

hazzlim commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -975,11 +978,74 @@ void VPlanTransforms::simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy) {
+    auto *NewBTC = new VPWidenCastRecipe(
+        Instruction::Trunc, Plan.getOrCreateBackedgeTakenCount(), NewIVTy);
+    Plan.getVectorPreheader()->appendRecipe(NewBTC);
+    auto *Cmp = dyn_cast<VPInstruction>(*WideIV->user_begin());
```
Good point - done.

hazzlim commented:
> One more thing I am wondering if we could end up introducing illegal vector types for a target and if we have to check the new type is legal?

I think that it should be fine, given that if the larger type was legal a smaller type with a power-of-2 size should be as well because it can always be promoted to a larger type if necessary. It's not obvious that we check the legality of using a vector of i64 (the default) as the induction variable in the first place?

fhahn reviewed (commented):
> > One more thing I am wondering if we could end up introducing illegal vector types for a target and if we have to check the new type is legal?
> 
> I think that it should be fine, given that if the larger type was legal a smaller type with a power-of-2 size should be as well because it can always be promoted to a larger type if necessary. It's not obvious that we check the legality of using a vector of i64 (the default) as the induction variable in the first place?

Yeah it should be fine I think.

One more thing to double check: interleave count doens't have to be a power of 2, could you add some tests where the IC isn't a power of 2 (e.g.3, 5) to make sure the rounding works as expected even if IC * VF isn't a power-of-2?

hazzlim commented:
> > > One more thing I am wondering if we could end up introducing illegal vector types for a target and if we have to check the new type is legal?
> > 
> > 
> > I think that it should be fine, given that if the larger type was legal a smaller type with a power-of-2 size should be as well because it can always be promoted to a larger type if necessary. It's not obvious that we check the legality of using a vector of i64 (the default) as the induction variable in the first place?
> 
> Yeah it should be fine I think.
> 
> One more thing to double check: interleave count doens't have to be a power of 2, could you add some tests where the IC isn't a power of 2 (e.g.3, 5) to make sure the rounding works as expected even if IC * VF isn't a power-of-2?

Hmm - it's actually quite hard to get this to trigger in a test. For one thing, as discussed above this patch will not trigger when we are interleaving https://github.com/llvm/llvm-project/pull/118828#discussion_r1890350878 so there's currently not much to test when interleaving.

As well as this, I think that the main use-case for this optimization is when we are tail folding and the Vector IV of interest is the mask - however if IC * VF isn't a power of 2 in that case we will hit the assert here:
https://github.com/llvm/llvm-project/blob/72768d9bb8ad3e97a852270726f04d7167d9ef50/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp#L2470-L2471

Do you think it would be ok to merge as-is and leave supporting interleaving as a follow-up?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -975,11 +978,78 @@ void VPlanTransforms::simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy) {
+    return false;
+
+  auto *TC = dyn_cast_if_present<ConstantInt>(
+      Plan.getTripCount()->getUnderlyingValue());
```
The underlying value should only be used if TripCount is a live-in (currently guaranteed I think), can you update this to use `getLiveInIRValue`, which will assert if it isn't a live-in?

fhahn replied:
Ah I see, the trip count is one of those special live-ins.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -975,11 +978,78 @@ void VPlanTransforms::simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy) {
+
+    // Currently only handle canonical IVs as it is trivial to replace the start
+    // and stop values, and we only perform the optimisation when the IV is only
+    // used by the comparison controlling loop control-flow.
```
Is the part about the comparison controlling loop control-flow accurate? We only check if it is used in a compare against the back edge-taken count, but this may not be used to control the loop control flow (which IIUC would mean exiting the loop?)

hazzlim replied:
Good point - I've simplified the comment to just say that we currently only consider IVs with a single use (the details of this being a comparison with BTC are documented in the comment below).

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -975,11 +978,78 @@ void VPlanTransforms::simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy) {
+    auto *WideIV = dyn_cast<VPWidenIntOrFpInductionRecipe>(&Phi);
+
+    // Currently only handle canonical IVs as it is trivial to replace the start
+    // and stop values, and we only perform the optimisation when the IV is only
```
```suggestion
    // and stop values, and we only perform the optimization when the IV is only
```

to be consistent with function comment & name?

hazzlim replied:
Done.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -975,11 +978,78 @@ void VPlanTransforms::simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy) {
+
+  VPBasicBlock *HeaderVPBB = Plan.getVectorLoopRegion()->getEntryBasicBlock();
+  for (VPRecipeBase &Phi : HeaderVPBB->phis()) {
+    auto *WideIV = dyn_cast<VPWidenIntOrFpInductionRecipe>(&Phi);
```
Can we also support VPWidenCanonicalIVPHIrecipe?

hazzlim replied:
I think we could, but the logic for changing the type that gets produced when the recipe is executed is slightly different to `VPWidenIntOrFpInductionRecipe`, so I was thinking to leave this to a follow up patch?

fhahn reviewed (commented):
> Hmm - it's actually quite hard to get this to trigger in a test. For one thing, as discussed above this patch will not trigger when we are interleaving [#118828 (comment)](https://github.com/llvm/llvm-project/pull/118828#discussion_r1890350878) so there's currently not much to test when interleaving.
> 
> As well as this, I think that the main use-case for this optimization is when we are tail folding and the Vector IV of interest is the mask - however if IC * VF isn't a power of 2 in that case we will hit the assert here:
> 
> https://github.com/llvm/llvm-project/blob/72768d9bb8ad3e97a852270726f04d7167d9ef50/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp#L2470-L2471
> 

I see, that sounds good. Left a few more comments.

hazzlim committed:
Address Review

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1805,6 +1805,9 @@ class VPWidenIntOrFpInductionRecipe : public VPWidenInductionRecipe {
              VPSlotTracker &SlotTracker) const override;
 #endif
 
+  /// Update the step value of the recipe.
+  void setStepValue(VPValue *V) { setOperand(1, V); }
+
```
can you move this to `VPWidenInductionRecipe`, after `getStepValue?

hazzlim replied:
Done.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.h:
```diff
@@ -97,6 +97,13 @@ struct VPlanTransforms {
                                  unsigned BestUF,
                                  PredicatedScalarEvolution &PSE);
 
+  /// Try to simplify the branch condition of \p Plan. This may restrict the
+  /// resulting plan to \p BestVF and \p BestUF.
+  static bool simplifyBranchConditionForVFAndUF(VPlan &Plan,
+                                                ElementCount BestVF,
+                                                unsigned BestUF,
+                                                PredicatedScalarEvolution &PSE);
+
```
This is only used in `optimizeForVFAndUF`, so it shouldn't need exposing here IIUC? It could be a static function in VPlanTransforms.cpp.

hazzlim replied:
Ah good point - done.

fhahn reviewed (commented):
Thanks for the latest updates.

Could you update the PR to latest main to make sure all tests are passing?

hazzlim committed:
Address review comments

- Move `setStepValue` to `VPWidenInductionRecipe`
- Make `simplifyBranchConditionForVFAndUF` a static, non-member function
- Update a lit test

hazzlim commented:
> Thanks for the latest updates.
> 
> Could you update the PR to latest main to make sure all tests are passing?

Thanks for the review - I have addressed the latest comments & updated to latest main, all tests seem to be passing with a minor update.

fhahn reviewed (approved):
LGTM, thanks

hazzlim committed:
Fix pattern matching for comparison with BTC

Following Commit 870f753 a broadcast is materialized for the BTC, so
update our pattern-matching logic to look for a comparison with a
broadcast of the BTC rather than the BTC itself.

hazzlim commented:
Ah - we needed a minor update to the pattern matching after https://github.com/llvm/llvm-project/commit/870f753f1f5e67aebccb46b158083a6c45bebce6, I've added https://github.com/llvm/llvm-project/pull/118828/commits/9906489a94889ff624a66ee47fa28a3d4a16e154 to fix this.

@fhahn please let me know if you're still happy with the patch :)

david-arm commented:
> Ah - we needed a minor update to the pattern matching after [870f753](https://github.com/llvm/llvm-project/commit/870f753f1f5e67aebccb46b158083a6c45bebce6), I've added [9906489](https://github.com/llvm/llvm-project/commit/9906489a94889ff624a66ee47fa28a3d4a16e154) to fix this.
> 
> @fhahn please let me know if you're still happy with the patch :)

The new pattern matching looks good to me!

fhahn commented on llvm/test/Transforms/LoopVectorize/tail-folding-optimize-vector-induction-width.ll:
```diff
@@ -0,0 +1,1006 @@
+}
+
+;.
+; CHECK: [[LOOP0]] = distinct !{[[LOOP0]], [[META1:![0-9]+]], [[META2:![0-9]+]]}
+; CHECK: [[META1]] = !{!"llvm.loop.isvectorized", i32 1}
+; CHECK: [[META2]] = !{!"llvm.loop.unroll.runtime.disable"}
+; CHECK: [[LOOP3]] = distinct !{[[LOOP3]], [[META2]], [[META1]]}
+; CHECK: [[LOOP4]] = distinct !{[[LOOP4]], [[META1]], [[META2]]}
+; CHECK: [[LOOP5]] = distinct !{[[LOOP5]], [[META2]], [[META1]]}
+; CHECK: [[LOOP6]] = distinct !{[[LOOP6]], [[META1]], [[META2]]}
+; CHECK: [[LOOP7]] = distinct !{[[LOOP7]], [[META2]], [[META1]]}
+; CHECK: [[LOOP8]] = distinct !{[[LOOP8]], [[META1]], [[META2]]}
+; CHECK: [[LOOP9]] = distinct !{[[LOOP9]], [[META2]], [[META1]]}
+; CHECK: [[LOOP10]] = distinct !{[[LOOP10]], [[META1]], [[META2]]}
+; CHECK: [[LOOP11]] = distinct !{[[LOOP11]], [[META2]], [[META1]]}
+; CHECK: [[LOOP12]] = distinct !{[[LOOP12]], [[META1]], [[META2]]}
+; CHECK: [[LOOP13]] = distinct !{[[LOOP13]], [[META2]], [[META1]]}
+; CHECK: [[LOOP14]] = distinct !{[[LOOP14]], [[META1]], [[META2]]}
+; CHECK: [[LOOP15]] = distinct !{[[LOOP15]], [[META2]], [[META1]]}
+; CHECK: [[LOOP16]] = distinct !{[[LOOP16]], [[META1]], [[META2]]}
+; CHECK: [[LOOP17]] = distinct !{[[LOOP17]], [[META2]], [[META1]]}
+; CHECK: [[LOOP18]] = distinct !{[[LOOP18]], [[META1]], [[META2]]}
+; CHECK: [[LOOP19]] = distinct !{[[LOOP19]], [[META2]], [[META1]]}
+;.
```
Can generate with `-check-globals=none` to avoid generating all of those.

hazzlim replied:
Done

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1009,11 +1012,83 @@ void VPlanTransforms::simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy) {
+  if (!TC || !BestVF.isFixed())
+    return false;
+
+  // Calculate the widest type required for known TC, VF and UF.
```
```suggestion
  // Calculate the minimum power-of-2 bit width that can fit the known TC, VF and UF. Returns at least 8.
```

hazzlim replied:
Ah thanks, that's a much clearer comment - done.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1063,7 +1138,7 @@ void VPlanTransforms::optimizeForVFAndUF(VPlan &Plan, ElementCount BestVF,
     VPBlockUtils::connectBlocks(Preheader, Header);
     VPBlockUtils::connectBlocks(ExitingVPBB, Exit);
-    simplifyRecipes(Plan, *CanIVTy);
+    VPlanTransforms::simplifyRecipes(Plan, *CanIVTy);
```
unrelated?

hazzlim replied:
This is actually necessary - because this logic has been moved into a (static) non-member function we now need the class scope when calling the `simplifyRecipes()`

fhahn commented on llvm/test/Transforms/LoopVectorize/tail-folding-optimize-vector-induction-width.ll:
```diff
@@ -0,0 +1,1006 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -p loop-vectorize -force-vector-width=8 -prefer-predicate-over-epilogue=predicate-dont-vectorize -S %s | FileCheck %s
```
Do we need VF=8 here or could we also choose 2, to get more compact IR?

hazzlim replied:
Ah good point - done.

fhahn reviewed (commented):
The pattern match looks fine to me. Spotted a few more remaining small things after another pass.

LGTM with those addressed if possible

hazzlim committed:
Address review comments

- Improve comment according to suggestion
- Generate test with `--check-globals=none`
- Use VF=2 for more compact IR in test

fhahn reviewed (approved):
LGTM, thanks!

hazzlim committed:
Merge branch 'main' into optimise-vector-indvar

hazzlim committed:
Fix formatting after merge

hazzlim commented:
> LGTM, thanks!

Thanks so much @fhahn and @david-arm for reviewing - I will wait for the buildbots to be happy after fixing the conflict with upstream/main and then I will land this.
